### PR TITLE
Detect circular dependencies during instance creation

### PR DESCRIPTION
**Pull request by BaronGreenback**
**#4709 Supercedes this PR.**

The 10.6 plugin i checked halted due to a di infinite loop.

This implements a list of DI types that have been asked to be created. If the list already contains one, a DI loop exists, and an exception is thrown, in an attempt to fail the original DI plugin load.

Works with the LDAP 9.0 and JF 10.7 RC 1

https://github.com/jellyfin/jellyfin/issues/4690

This is not a fix - it does not make the plugins work - it just attempts to stop outdated plugins from hanging the system.

It is a stop gap method until a better solution can be implemented.

### CODE REVIEW

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 1]
```diff
@@ -393,8 +393,20 @@ public T CreateInstance<T>()
...
         {
+            if (_creatingInstances == null)
+            {
+                _creatingInstances = new List<Type>();
+            }
+
+            if (_creatingInstances.IndexOf(type) != -1)
+            {
+                Logger.LogError("Incompatible plugin detected. Attempting to bypass.");
+                throw new ExternalException("Incompatible plugin.");
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs:404]: Doesn't have to be a plugin and also the function is called `CreateInstanceSafe`. It's no longer safe ;)

Comment (BaronGreenback) [reply Emby.Server.Implementations/ApplicationHost.cs:404]: Now called CreateInstanceWithSafeguards

Naff name, but i couldn't think of better name besides CreateInstanceInASlightlyMoreSecureWayButItStillMightFailAndDontPutItInAnExceptionHandler.

**Comment by joshuaboniface** [comment]
>This is not a fix - it does not make the plugins work - it just attempts to stop outdated plugins from hanging the system.

This is how it should work, so that's fine. If a plugin can't be loaded, it's just dropped. A toast/notification of some kind in this case would be ideal too, but probably outside of bugfix scope - ignoring the failing plugin is how 10.6.z worked.

**Comment by BaronGreenback** [comment]
> > This is not a fix - it does not make the plugins work - it just attempts to stop outdated plugins from hanging the system.
> 
> This is how it should work, so that's fine. If a plugin can't be loaded, it's just dropped. A toast/notification of some kind in this case would be ideal too, but probably outside of bugfix scope - ignoring the failing plugin is how 10.6.z worked.

@joshuaboniface #4709 handles plugins much more effectively, and should solve most of the issues we are having. Am in the process of testing now.

**Comment by cvium** [comment]
The thing is that what is changed here is a very central method. It's not just plugins that are affected.
And it's not just dropped, the server will stop. There's no way to recover without removing the offending classes/assemblies.

**Comment by BaronGreenback** [comment]
v8 and v9 of the LDAP plugin both caused a DI loop, which this does detect - and permits access to the system. Given the amount of changes within the core, the chances are that this will occur again in the future, as code gets moved around.

It isn't a fix - as said above - it's an attempt to get JF to load after an upgrade. 

The ideal solution, is to include the auto-updating of plugins as part of the installation "wizard".

**Comment by crobibero** [comment]
Closed in favor of https://github.com/jellyfin/jellyfin/pull/4709/

**Pull request closed without merging**